File: Python/_0054_Spiral_Matrix/Solution.py
# coding=utf-8
from typing import List


# 执行用时：40 ms, 在所有 Python3 提交中击败了60.05%的用户
# 内存消耗：14.8 MB, 在所有 Python3 提交中击败了80.38%的用户
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n = len(matrix); m = len(matrix[0])
        dx = [0, 1, 0, -1]; dy = [1, 0, -1, 0]
        st = [[False for _ in range(m)] for _ in range(n)]

        res = []
        x = y = d = 0
        for i in range(n * m):
            res.append(matrix[x][y])
            st[x][y] = True
            a = x + dx[d]; b = y + dy[d]
            if a < 0 or a >= n or b < 0 or b >= m or st[a][b]:
                d = (d + 1) % 4
                a = x + dx[d]; b = y + dy[d]
            x = a; y = b
        return res


if __name__ == "__main__":
    matrix = [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]
    ]
    print(Solution().spiralOrder(matrix))


# class Solution(object):
#     def spiralOrder(self, matrix):
#         """
#         :type matrix: List[List[int]]
#         :rtype: List[int]
#         """
#         res = []
#
#         while matrix:
#             res.extend(matrix.pop(0))
#             matrix = list(map(list, zip(*matrix)))[::-1]
#             print(matrix)
#             if not matrix or not matrix[0]:
#                 break
#         return res

File: Java/src/_0054_Spiral_Matrix/Solution.java
package _0054_Spiral_Matrix;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 时间复杂度：O(N*M)
 * 空间复杂度：O(N*M)
 * 执行用时：0 ms, 在所有 Java 提交中击败了100.00%的用户
 * 内存消耗：36.8 MB, 在所有 Java 提交中击败了8.33%的用户
 */
class Solution {

    public int[][] d = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    public int N, M;  // 行数，列数

    public List<Integer> spiralOrder(int[][] matrix) {

        N = matrix.length;
        if (N == 0) return new ArrayList<>();
        M = matrix[0].length;
        if (M == 0) return new ArrayList<>();

        boolean[][] st = new boolean[N][M];
        for (int i = 0; i < N; i++) Arrays.fill(st[i], false);

        ArrayList<Integer> res = new ArrayList<>();
        int dir = 0, x = 0, y = 0;
        while (res.size() < N * M) {
            if (!st[x][y]) {
                res.add(matrix[x][y]);
                st[x][y] = true;
            }
            int a = x + d[dir][0], b = y + d[dir][1];
            if (a >= 0 && a < N && b >= 0 && b < M && !st[a][b]) {
                x = a;
                y = b;
            } else dir = (dir + 1) % 4;
        }
        return res;
    }

    public static void main(String[] args) {
        int[][] matrix = {
                {1, 2, 3},
                {4, 5, 6},
                {7, 8, 9}
        };
        System.out.println((new Solution()).spiralOrder(matrix));
    }
}
